arangod/VocBase/Methods/Collections.cpp

```diff
@@ -289,18 +289,50 @@ Result Collections::create(TRI_vocbase_t& vocbase,
     helper.add(arangodb::StaticStrings::DataSourceName,
                arangodb::velocypack::Value(info.name));
 
-    if (ServerState::instance()->isCoordinator()) {
+    // If the PlanId is not set, we either are on a single server, or this is
+    // a local collection in a cluster; which means, it is neither a user-facing
+    // collection (as seen on a Coordinator), nor a shard (on a DBServer).
+    bool const isLocalCollection =
+        (!ServerState::instance()->isCoordinator() &&
+         Helper::stringUInt64(info.properties.get(StaticStrings::DataSourcePlanId)) == 0);
+
+    bool const isSystemName = TRI_vocbase_t::IsSystemName(info.name);
+
+    // All collections on a single server should be local collections.
+    // A Coordinator should never have local collections.
+    // On an Agent, all collections should be both local and system collections.
+    // On a DBServer, the only local collections should be system collections
+    // (like _statisticsRaw). Non-local (system or not) collections are shards,
+    // so don't have system-names, even if they are system collections!
+    switch(ServerState::instance()->getRole()) {
+      case ServerState::ROLE_SINGLE:
+        TRI_ASSERT(isLocalCollection);
+        break;
+      case ServerState::ROLE_DBSERVER:
+        TRI_ASSERT(isLocalCollection == isSystemName);
+        break;
+      case ServerState::ROLE_COORDINATOR:
+        TRI_ASSERT(!isLocalCollection);
+        break;
+      case ServerState::ROLE_AGENT:
+        TRI_ASSERT(isLocalCollection && isSystemName);
```

Comment:
Is this going to work? I think an agent has collections named `log` and `compact` and `config`, which are not system collections. They will be local, but will not have a system name. They are created only once pretty early on, maybe it works for this reason.

Reply:
Ah. I wrongly assumed that the agents' collections would start with an underscore.

The tests are successful, but I better fix that assertion anyway. Thanks!

Reply:
I've fixed the assertion. And i also verified that this code is just not executed on agents to create those collections. They are created in `arangod/Agency/State.cpp`, in `State::createCollections()`, which directly calls `TRI_vocbase_t::createCollection()`.